bucketers.py

```diff
@@ -156,6 +156,37 @@ def calc_source_bucketed_matches(self, src, ref, out, ref_aligns, out_aligns, sr
         fmeas = 2 * prec * rec / (prec + rec)
       yield both_tot, ref_tot, out_tot, rec, prec, fmeas
 
+  def calc_bucketed_likelihoods(self, corpus, likelihoods):
+    """
+    Calculate the average of log likelihoods, bucketed by the type of word/label we have
+    This must be used with a subclass that has self.bucket_strs defined, and self.calc_bucket(word) implemented.
+
+    Args:
+      corpus: The text/label corpus over which we compute the likelihoods
+      likelihoods: The log-likelihoods corresponding to each word/label in the corpus
+
+    Returns:
+      the average log-likelihood bucketed by the type of word/label we have
+    """
+
+    if type(corpus) == str:
+      corpus = corpus_utils.load_tokens(corpus)
+    bucketed_likelihoods = [[0.0, 0] for _ in self.bucket_strs]
+    assert (len(corpus) == len(likelihoods))
+    for sent, list_of_likelihoods in zip(corpus, likelihoods):
+      assert (len(sent) == len(list_of_likelihoods))
+      for word, ll in zip(sent, list_of_likelihoods):
+        bucket = self.calc_bucket(word, ref_label=word)
+        bucketed_likelihoods[bucket][0] += ll
+        bucketed_likelihoods[bucket][1] += 1
+
+    for ll, count in bucketed_likelihoods:
+      if count != 0:
+        yield ll/float(count)
+      else:
+        yield -float("inf") # large negative value
```

Comment:
It's not actually `-inf`, but rather "NA", correct? I'm mainly suggesting making this clear because in other cases it could legitimately be `-inf` when you do have counts (e.g. when there's a word that is not included in the vocabulary of a closed-vocabulary model).